File: 143_Reorder List.java
// Time complexity: O(n)
// Space complexity: O(1)

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public void reorderList(ListNode head) {
        // 1. LC 876: find the middle of the list: slow
        ListNode fast = head, slow = head;
        while (fast != null && fast.next != null) {
            fast = fast.next.next;
            slow = slow.next;
        }
        // 2. LC 206: reverse the list starting from middle: new_reversed_head: prev
        ListNode cur = slow, prev = null, tmp;
        while (cur != null) {
            tmp = cur.next;
            cur.next = prev;
            prev = cur;
            cur = tmp;
        }
        // 3. repeatedly collect nodes from head->middle-1 & reverse tail->middle
        // ListNode vHead = new ListNode();
        ListNode tmp1, tmp2;
        // while (head != null && prev != null && head != prev && head.next != prev) {
        while (prev.next != null) {
            tmp1 = head.next;
            tmp2 = prev.next;
            head.next = prev;
            prev.next = tmp1;
            head = tmp1;
            prev = tmp2;
        }
    }
}

File: 143_Reorder List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        # 1. find median (the last of left half)
        fast, slow = head, head
        while fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next
        
        right = slow.next
        
        # 2. break the connection between left and right
        slow.next = None
        left = head
        
        # 3. reverse right
        def reverse(node) -> Optional[ListNode]:
            vHead = ListNode()
            while node:
                tmp = node.next
                node.next = vHead.next
                vHead.next = node
                node = tmp
            return vHead.next
        
        right = reverse(right)
        
        # 4. append l/r repeatedly - end case: invalid r as l always appended to the cur valid tail already
        while right:
            n1, n2 = left.next, right.next
            left.next = right
            right.next = n1
            left, right = n1, n2